### PR TITLE
fault tolerance

### PR DESCRIPTION
**Pull request by NeGnail**
It is common to have multiple service providers for a service in a cluster. Some of these service providers may have a network, configuration, a long fullgc, a full thread pool, a hardware failure, etc. The long connection is alive but the program is not responding properly. This functionality demotes the service provider for this part of the exception. This feature solves the problem of service failures that continue to affect the business and avoids the avalanche effect. The long response time required by manual intervention can be reduced and the system availability can be improved.

The core logic of this function is referred to from [sofa-rpc](https://www.sofastack.tech/projects/sofa-rpc/fault-tolerance/), which has been used in many large-scale scenarios with good results.

### CODE REVIEW

File: pkg/filter/stream/faulttolerance/regulator/invocation_dimension_factory.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"mosn.io/api"
+)
+
+type NewDimension func(info api.RequestInfo) InvocationDimension
+
+var dimensionNewFunc NewDimension
+
+func RegisterNewDimension(newDimension NewDimension) {
```
[the remark this answers is not in the document]
Comment (NeGnail) [reply pkg/filter/stream/faulttolerance/regulator/invocation_dimension_factory.go:11]: this is left to the consumer to implement

File: pkg/filter/stream/faulttolerance/regulator/default_work_pool.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	index := w.random()
+	if value, ok := w.workers.Load(index); ok {
+		worker := value.(*WorkGoroutine)
+		worker.AddTask(model.GetKey(), model)
+	} else {
+		worker := NewWorkGoroutine()
+		worker.AddTask(model.GetKey(), model)
+		if _, ok := w.workers.LoadOrStore(index, worker); !ok {
+			worker.Start()
+			w.workSize++
```
[the remark this answers is not in the document]
Comment (NeGnail) [reply pkg/filter/stream/faulttolerance/regulator/default_work_pool.go:84]: is used by DefaultWorkPool.random()

File: pkg/filter/stream/faulttolerance/regulator/invocation_stat_factory.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+var invocationStatFactoryInstance *InvocationStatFactory
+
+func GetInvocationStatFactoryInstance() *InvocationStatFactory {
+	return invocationStatFactoryInstance
+}
+
+func NewInvocationStatFactory(config *v2.FaultToleranceFilterConfig) *InvocationStatFactory {
+	invocationStatFactory := &InvocationStatFactory{
```
Comment (wangfakang) [inline pkg/filter/stream/faulttolerance/regulator/invocation_stat_factory.go:21]: should use `invocationStatFactoryInstance`.

File: pkg/filter/stream/faulttolerance/regulator/measure_model.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+	snapshots := []*InvocationStat{}
+	m.stats.Range(func(app, value interface{}) bool {
+		stat := value.(*InvocationStat)
+		if !stat.IsHealthy() {
+			m.recover(stat, recoverTime)
+			return true
+		}
+		if stat.GetCall() <= 0 {
+			if stat.AddUselessCycle() {
+				m.releaseInvocationStat(stat)
```
Comment (wangfakang) [inline pkg/filter/stream/faulttolerance/regulator/measure_model.go:95]: Maybe caused to unhealth stat never recover.

File: pkg/filter/stream/faulttolerance/regulator/invocation_stat_factory.go (hunk 3 above)
Comment (wangfakang) [reply pkg/filter/stream/faulttolerance/regulator/invocation_stat_factory.go]: Done, https://github.com/NeGnail/sofa-mosn/pull/2

File: pkg/filter/stream/faulttolerance/regulator/invocation_dimension_factory.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (wangfakang) [reply pkg/filter/stream/faulttolerance/regulator/invocation_dimension_factory.go:16]: Done，https://github.com/NeGnail/sofa-mosn/pull/2

File: pkg/config/v2/filter.go [hunk 5]
```diff
@@ -47,6 +47,18 @@ const (
 	ORIGINALDST_LISTENER_FILTER = "original_dst"
 )
 
+type FaultToleranceFilterConfig struct {
+	Enabled               bool `json:"enabled"`
+	ExceptionTypes        map[uint32]bool
```
Comment (nejisama) [inline pkg/config/v2/filter.go:52]: json格式注释没有写全

File: pkg/filter/stream/faulttolerance/regulator/default_regulator.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package regulator
+
+import (
+	"mosn.io/mosn/pkg/config/v2"
```
Comment (nejisama) [inline pkg/filter/stream/faulttolerance/regulator/default_regulator.go:4]: goimports

File: pkg/filter/stream/faulttolerance/regulator/invocation_dimension_factory.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	return d.measureKey
+}
+
+func defaultDimensionFunc(info api.RequestInfo) InvocationDimension {
+	if info == nil || info.UpstreamHost() == nil || info.RouteEntry() == nil {
+		return &defaultDimension{}
+	} else {
+		return &defaultDimension{
+			invocationKey: info.UpstreamHost().AddressString(),
+			measureKey:    info.RouteEntry().ClusterName(),
```
Comment (nejisama) [inline pkg/filter/stream/faulttolerance/regulator/invocation_dimension_factory.go:60]: 这里最好记录一个TODO，严格来说Route.ClusterName的结果是可能存在变化的。虽然目前实际没有用到这种场景

File: pkg/filter/stream/faulttolerance/regulator/invocation_stat.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+	return atomic.LoadUint64(&s.callCount), atomic.LoadUint64(&s.exceptionCount)
+}
+
+func (s *InvocationStat) GetDowngradeTime() int64 {
+	return s.downgradeTime
+}
+
+func (s *InvocationStat) AddUselessCycle() bool {
+	s.uselessCycle++
+	return s.uselessCycle >= 6
```
Comment (nejisama) [inline pkg/filter/stream/faulttolerance/regulator/invocation_stat.go:78]: 这个6 是什么magic number

File: pkg/filter/stream/faulttolerance/regulator/invocation_stat.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+	"fmt"
+	"sync/atomic"
+
+	"mosn.io/api"
+	"mosn.io/mosn/pkg/filter/stream/faulttolerance/util"
+)
+
+type InvocationStat struct {
+	dimension      InvocationDimension
+	host           *api.HostInfo
```
Comment (nejisama) [inline pkg/filter/stream/faulttolerance/regulator/invocation_stat.go:30]: api.HostInfo是一个interface，这里设计成它的指针是不是误以为是一个结构体了？

File: pkg/filter/stream/faulttolerance/regulator/measure_model.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	m.updateInvocationSnapshots(snapshots)
+}
+
+func (m *MeasureModel) downgrade(snapshot *InvocationStat, maxIpCount uint64) {
+	if atomic.LoadUint64(&m.count) <= 0 {
+		return
+	}
+	if m.downgradeCount+1 > maxIpCount {
```
Comment (nejisama) [inline pkg/filter/stream/faulttolerance/regulator/measure_model.go:93]: 为什么部分有并发控制（atomic），部分没有？

File: pkg/filter/stream/faulttolerance/send_filter.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+type SendFilter struct {
+	config            *v2.FaultToleranceFilterConfig
+	invocationFactory *regulator.InvocationStatFactory
+	handler           api.StreamSenderFilterHandler
+}
+
+func NewSendFilter(config *v2.FaultToleranceFilterConfig) *SendFilter {
+	filter := &SendFilter{
+		config:            config,
+		invocationFactory: regulator.NewInvocationStatFactory(config),
```
Comment (nejisama) [inline pkg/filter/stream/faulttolerance/send_filter.go:38]: 这个Filter应该是每个请求都会生成一个，每个请求都生成一个这个StatFactory的逻辑看上去似乎不太对

File: pkg/filter/stream/faulttolerance/regulator/invocation_stat.go (hunk 8 above)
Comment (NeGnail) [reply pkg/filter/stream/faulttolerance/regulator/invocation_stat.go:78]: 最大没有调用的计算轮次。改为一个常量了。

File: pkg/filter/stream/faulttolerance/regulator/measure_model.go (hunk 10 above)
Comment (NeGnail) [reply pkg/filter/stream/faulttolerance/regulator/measure_model.go:93]: 如果只会有一个协程在跑的就不加并发控制了。

File: pkg/filter/stream/faulttolerance/regulator/invocation_stat_factory.go [hunk 12]
```diff
@@ -30,20 +30,24 @@ type InvocationStatFactory struct {
 }
 
 var invocationStatFactoryInstance *InvocationStatFactory
+var onceStatFactoryInstance = new(sync.Once)
 
-func GetInvocationStatFactoryInstance() *InvocationStatFactory {
+func GetInvocationStatFactoryInstance(config *v2.FaultToleranceFilterConfig) *InvocationStatFactory {
```
Comment (wangfakang) [inline pkg/filter/stream/faulttolerance/regulator/invocation_stat_factory.go:35]: the `once` maybe causes the config not to be dynamically updated.

**Pull request merged**